ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/CompletionDeduplicationInfoIT.scala

```diff
@@ -0,0 +1,182 @@
+    val expectedDeduplicationTime = Some(requestedDeduplicationTime)
+    val actualDeduplicationTime = completion.deduplicationPeriod.deduplicationTime
+    assert(completion.status.forall(_.code == Status.Code.OK.value()))
+    assert(
+      actualDeduplicationTime == expectedDeduplicationTime,
+      "Wrong duplication time in completion, " +
+        s"expected: $expectedDeduplicationTime, " +
+        s"actual: $actualDeduplicationTime",
+    )
+  }
+
+  private def assertSubmissionIdIsPreserved(
+      requestedSubmissionId: Ref.SubmissionId,
+      optCompletion: Option[Completion],
+  ): Unit = {
+    val submissionIdCompletion = assertDefined(optCompletion)
+    val actualSubmissionId = submissionIdCompletion.submissionId
+    assert(submissionIdCompletion.status.forall(_.code == Status.Code.OK.value()))
+    assert(
+      actualSubmissionId == requestedSubmissionId,
+      "Wrong submission ID in completion, " +
+        s"expected: $requestedSubmissionId, actual: $actualSubmissionId",
+    )
+  }
+
+  private def assertDefaultDeduplicationTimeIsReported(
+      config: LedgerConfiguration,
+      optCompletion: Option[Completion],
+  ): Unit = {
+    val completion = assertDefined(optCompletion)
+    assert(completion.status.forall(_.code == Status.Code.OK.value()))
+    val actualDeduplication = completion.deduplicationPeriod
+    assert(actualDeduplication.isDeduplicationTime)
+    assert(
+      actualDeduplication.deduplicationTime == config.maxDeduplicationTime,
+      s"The deduplication $actualDeduplication " +
+        "is not the maximum deduplication time",
+    )
+  }
+
+  private def assertSubmissionIdIsGenerated(optCompletion: Option[Completion]): Unit = {
+    val completion = assertDefined(optCompletion)
+    assert(completion.status.forall(_.code == Status.Code.OK.value()))
+    assert(
+      Ref.SubmissionId.fromString(completion.submissionId).isRight,
+      "Missing or invalid submission ID in completion",
+    )
+  }
+
+  private def assertApplicationIdIsPreserved(
+      requestedApplicationId: String,
+      optCompletion: Option[Completion],
+  ): Unit = {
+    val expectedApplicationId = requestedApplicationId
+    assertDefined(optCompletion)
+    val applicationIdCompletion = optCompletion.get
+    assert(applicationIdCompletion.status.forall(_.code == Status.Code.OK.value()))
+    val actualApplicationId = applicationIdCompletion.applicationId
+    assert(
+      Ref.ApplicationId.fromString(actualApplicationId).contains(expectedApplicationId),
+      "Wrong application ID in completion, " +
+        s"expected: $expectedApplicationId, actual: $actualApplicationId",
+    )
+  }
+
+  private def assertDefined(optCompletion: Option[Completion]): Completion = {
+    assert(optCompletion.isDefined, "No completion has been produced")
+    optCompletion.get
+  }
+
+  private def simpleCreate(party: Primitive.Party): Command = Dummy(party).create.command
+
+  private val aSubmissionId =
```

Comment:
Minor: I would put `val`s above `def`s.

Reply:
Why? I like top-bottom code w.r.t. LOD and I see these constants as quite irrelevant details.

Reply:
If it was a class, it would fail, I think, with a forward reference.  However, it makes sense to keep them at the bottom, ignore.

Reply:
Even in that case, I'd rather use `lazy`.